documents/write_acls.go

```diff
@@ -0,0 +1,86 @@
+package documents
+
+import (
+	"bytes"
+
+	"github.com/centrifuge/precise-proofs/proofs"
+)
+
+// changedField holds the compact property, old and new value of the field that is changed
+// if the old is nil, then it is a set operation
+// if new is nil, then it is an unset operation
+// if both old and new are set, then it is an edit operation
+type changedField struct {
+	property, old, new []byte
+}
+
+// getChangedFields takes two document trees and returns the compact value, old and new value of the fields that are changed in new tree.
+// Properties may have been added to the new tree or removed from the new tree.
+// In Either case, since the new tree is different from old, that is considered a change.
+func getChangedFields(oldTree, newTree *proofs.DocumentTree, lengthSuffix string) (changedFields []changedField) {
+	oldProps := oldTree.PropertyOrder()
+	newProps := newTree.PropertyOrder()
+
+	props := make(map[string]proofs.Property)
+	for _, p := range append(oldProps, newProps...) {
+		// we can ignore the length property since any change in slice or map will return in addition or deletion of properties in the new tree
+		if p.Text == lengthSuffix {
+			continue
+		}
+
+		if _, ok := props[p.ReadableName()]; ok {
+			continue
+		}
+
+		props[p.ReadableName()] = p
+	}
+
+	// check each property and append it changed fields if the value is different.
+	for _, p := range props {
+		_, ol := oldTree.GetLeafByProperty(p.ReadableName())
+		_, nl := newTree.GetLeafByProperty(p.ReadableName())
+
+		if ol == nil {
+			changedFields = append(changedFields, newChangedField(p, nl, false))
```

Comment:
shouldn't this append be done only if the other field new or old is not nil?

Reply:
in my understanding, old == nil in this case means that the field has been added in the new version of the doc, so the fields needs to be appended to the list of changed fields.

Reply:
So, if the property doesn't exist in old one but exists in new one, that means there is an addition operation in the new document. If the property is not found in the new document but found in the old document, then there was a deletion operation in the new document.

Reply:
The second for-loop is off putting to me, thats why I got confused, why do we need it? can't we just iterate over the entire list of fields and do skip same ones using the map while processing the changes?

Reply:
lets talk offline, not sure I understand your question,

Reply:
I meant removing the second for loop, I hope its obvious from what I said, anyway we can always discuss. I can also do the refactoring.